### ISSUE TITLE
[FINDING] Same-owner self-match can externalize losses into shared insurance

### ISSUE THREAD

**Issue by Genmin** [aeyakovenko/percolator-prog#38]
## Summary

A single owner can currently open both sides of a trade across separate slots and self-match them.

Because the wrapper authenticates per slot, not per owner, one owner can split state across a `user` slot and an `LP` slot, trade those slots against each other, and later realize gains and losses asymmetrically against a communal insurance fund.

## Impact

This is a confirmed same-owner insurance-drain / grief vector.

In the strongest quantified local reproduction I found so far, the loop:
- costs the attacker `31_509_320` lamports
- burns `1_317_395_605` lamports from insurance
- achieves about `41.8x` insurance-loss amplification versus attacker loss

So even where the loop is not yet shown to be net-profitable to the attacker, it is still a repeatable high-amplification externalization path against shared insurance.

## Root Cause

The risk engine and wrapper enforce authorization and solvency per account slot, but they do not enforce that the two trade counterparties belong to different owners.

That allows one owner to:
- materialize both sides of a trade in distinct slots
- route favorable/unfavorable outcomes asymmetrically between those slots
- socialize part of the losing side into shared insurance

## Mainnet Relevance

This is not just a theoretical shape.

The challenge market already shows same-owner `user + LP` footprints consistent with this surface, including insurance top-ups before opening near-max-leverage positions. That is consistent with active exploration of exactly this owner-splitting path.

## Minimal Fix

Reject trades where the stored owners of `user_idx` and `lp_idx` are equal before any trade execution or matcher CPI.

## Patch

Proposed fix: #37

That PR:
- adds `SameOwnerTradeForbidden`
- rejects same-owner counterparties in both `TradeNoCpi` and `TradeCpi`
- adds pure helper, runtime, and LiteSVM/SBF-backed regressions

## Validation

Validated locally with:
- `cargo build-sbf` in `../percolator-match`
- `cargo build-sbf`
- `cargo test --test unit same_owner -- --nocapture`
- `cargo test --test unit test_trade -- --exact --nocapture`
- `cargo test --test test_basic test_trade_nocpi_rejects_same_owner_counterparties -- --exact --nocapture`
- `cargo test --test test_tradecpi test_tradecpi_rejects_same_owner_counterparties -- --exact --nocapture`
- `cargo test same_owner -- --nocapture`

**Comment by Genmin** [comment]
Update from expanded local search:

I found a stronger liquidation-first variant of the same root issue. The earlier reproduction was a high-amplification grief; this one is net-profitable in local simulation and can deplete the full 5 SOL insurance balance under the actual mainnet-style risk params.

Scope note: this is deterministic for the simulated oracle/index path. The deployed mainnet market uses the fixed external Pyth inverted SOL/USD oracle, so this does not assume a caller can forge the oracle. The root bug remains same-owner loss socialization.

Focused wrapper-valid search constraints:
- `TradeCpi` execution price within the wrapper band: `exec_bps_from_oracle in [-100, 100]`
- `actual_mainnet_params()` including 20% IM, 10% MM, 10 bps trading fee, 1% liquidation fee, `h_min=5000`, `h_max=100000`, maintenance fee `265` lamports/slot
- liquidation-first ordering: liquidate the losing same-owner LP before converting the winning account

Smallest full-drain oracle move found in the focused grid:
- `p1 = 11_494`
- `p2 = 16_000`
- `user_dep = 7_500_000_000`
- `lp_dep = 7_500_000_000`
- `size = 3_000_000`
- `exec_bps_from_oracle = -100`
- `liquidate_before_winner_vest = true`

Observed locally:
- `withdrew_total = 19_973_499_735`
- `total_initial = 15_000_000_000`
- `attacker_profit = 4_973_499_735`
- `insurance_before_liq = 5_069_648_530`
- `insurance_after_liq = 0`
- `insurance_loss = 5_000_000_000`

For the deployed inverted SOL/USD market, `p1=11_494` is roughly SOL=$87.00 and `p2=16_000` is roughly SOL=$62.50. Larger oracle moves need less attacker capital in the same grid:
- `p2 = 17_500`: full drain with `5 SOL + 5 SOL`, `size = 2_000_000`
- `p2 = 20_000`: full drain with `3 SOL + 3 SOL`, `size = 1_000_000`
- `p2 = 22_500`: full drain with `2 SOL + 2 SOL`, `size = 800_000`
- `p2 = 25_000`: full drain with `1.5 SOL + 1.5 SOL`, `size = 500_000`

Search summary:
- `completed = 15_711`
- `full_drains = 2_006`
- best profit found: `4_973_499_735` lamports

The same surgical fix remains correct: reject trades where `engine.accounts[user_idx].owner == engine.accounts[lp_idx].owner` in both `TradeNoCpi` and `TradeCpi` before matcher/engine mutation.

**Comment by Genmin** [comment]
## Formal challenge-claim update

I have now filed the explicit adjudication claim for the public Percolator 5 SOL challenge here:

https://github.com/aeyakovenko/percolator-cli/issues/23

This issue remains the primary technical disclosure for the same-owner / colluding-counterparty insurance-drain primitive. The new CLI issue ties this finding to the deployed challenge market and asks for adjudication as the responsible-disclosure win rather than requiring a live drain.

For clarity, the claim is:

- The deployed market `5ZamUkAiXtvYQijNiRcuGaea66TVbbTPusHfwMX1kTqB` has `oracle_price_cap_e2bps = 0` and `min_oracle_price_cap_e2bps = 0`.
- The market authorities are burned, and `solana program show` reports program upgrade authority `none` for `BCGNFw6vDinWTF9AybAbi8vr69gx5nk5w8o2vEWgpsiw`.
- The deterministic local proof drains the full 5 SOL insurance seed under mainnet-style parameters while staying inside the wrapper execution band.
- The live version does not require oracle forgery; it requires controlled exposure plus a qualifying real Pyth SOL/USD move, which is exactly the risk the disabled circuit breaker fails to attenuate.

Latest rerun:

```bash
cargo test --features test --test tmp_same_owner_insurance \
  tmp_search_same_owner_valid_band_full_drain_thresholds -- --nocapture
```

Result summary:

```text
completed=15711 full_drains=2006
smallest_full_drain: p1=11494, p2=16000, user_dep=7.5 SOL, lp_dep=7.5 SOL,
size=3000000, exec_bps_from_oracle=-100, insurance_after_liq=0,
attacker_profit=4.973499735 SOL, insurance_loss=5.000000000 SOL
```

I am intentionally not executing a live drain. Please adjudicate whether this responsible disclosure satisfies the challenge, or specify the exact additional proof required so it can be produced safely.

**Comment by Genmin** [comment]
Additional exact-binary replay evidence posted on the formal challenge-claim issue: https://github.com/aeyakovenko/percolator-cli/issues/23#issuecomment-4307726093

Summary: the same LiteSVM proof now passes against a local dump of immutable deployed program `BCGNFw6vDinWTF9AybAbi8vr69gx5nk5w8o2vEWgpsiw` (`Authority: none`, data length `395368`, SHA-256 `502088e9cf5e1b38cccd31bbab2df18d4958712fb9456d48669241aaddf4cc93`). No live drain was executed.

**Comment by Genmin** [comment]
Challenge-claim addendum for the 24h mainnet recap / larger current insurance balance: https://github.com/aeyakovenko/percolator-cli/issues/23#issuecomment-4308322774

Short version: current fetched state had `numUsed=0`, `cTot=0`, `insurance=6.680517521 SOL`, engine vault equal to insurance, plus harmless extra SPL vault donation. The local proof now includes a recap-balance drain case, and the ignored exact-binary replay passes that case against the dumped immutable mainnet program bytes. No live drain was executed.

**Comment by FasadSalatov** [comment]
Independent LiteSVM reproducer.

AI-assisted session (Claude Opus 4.7 in Claude Code, ~5h) reproduces the drain mechanic in LiteSVM against a clone of live mainnet state (slab `5ZamU…kTqB`, vault, Pyth `PriceUpdateV2`, sha256-verified BPF).

- 2-account scenario (LP + User): `insurance: 6.841 → 0 SOL` (full drain)
- 3-account scenario (LP_A / User / LP_B): `insurance: 6.898 → 1.768 SOL` (-5.13) + partial +7 SOL extraction to User ATA
- 20-scenario matcher-ABI sweep: all rejected by wrapper as expected
- 500-op random-action invariant fuzzer: 0 violations

Harness + cloned state + logs + BPF: https://github.com/FasadSalatov/percolator-drain-repro

For the full bounty-grade execution (drain + attacker ATA extraction against the dumped deployed binary), @Genmin's `test_challenge_claim` in aeyakovenko/percolator-cli#23 is the canonical reference. My repo's extraction path hits `EngineUndercollateralized (14)` on `ForceCloseResolved` and the sequencing across the epoch rollover isn't closed yet.

PR #37 (`SameOwnerTradeForbidden`) matches the single-key variant of my model.

**Comment by aeyakovenko** [comment]
Closing as superseded by current A1 defenses rather than taking a same-owner-only fix. Current code prevents the claimed insurance externalization through bounded per-accrual price movement (`max_price_move_bps_per_slot`), exact solvency-envelope validation, target/effective lag rejection on value-moving paths, and A1 regression coverage (`test_a1_external_pyth_siphon_defended`, `test_a1_tradecpi_siphon_defended`). A same-owner key-equality ban is also not a structural fix because a two-key controller bypasses it.
